loader/loader.go

```diff
@@ -1188,110 +1188,245 @@ func fetchMatchedLiteral(ctx *tcontext.Context, router *router.Table, schema, ta
 	return targetSchema, targetTable
 }
 
-func (l *Loader) restoreData(ctx context.Context) error {
-	begin := time.Now()
+// `restore Schema Job` present a data structure of schema restoring job
+type restoreSchemaJob struct {
+	loader   *Loader
+	session  *DBConn
+	database string // database name
+	table    string // table name, empty if it's a schema of database
+	filepath string // file path of dumpped schema file
+}
 
-	baseConn, err := l.toDB.GetBaseConn(ctx)
-	if err != nil {
-		return err
+// `jobQueue` of schema restoring which (only) support consumptions concurrently
+type jobQueue struct {
+	ctx           context.Context
+	msgq          chan *restoreSchemaJob // job message queue channel
+	consumerCount int                    // count of consumers
+	eg            *errgroup.Group        // err wait group of consumer's go-routines
+}
+
+// `newJobQueue` consturct a jobQueue
+func newJobQueue(ctx context.Context, consumerCount, length int) *jobQueue {
+	eg, selfCtx := errgroup.WithContext(ctx)
+	return &jobQueue{
+		ctx:           selfCtx,
+		msgq:          make(chan *restoreSchemaJob, length),
+		consumerCount: consumerCount,
+		eg:            eg,
 	}
-	defer func() {
-		err2 := l.toDB.CloseBaseConn(baseConn)
-		if err2 != nil {
-			l.logger.Warn("fail to close connection", zap.Error(err2))
-		}
-	}()
+}
 
-	dbConn := &DBConn{
-		cfg:      l.cfg,
-		baseConn: baseConn,
-		resetBaseConnFn: func(*tcontext.Context, *conn.BaseConn) (*conn.BaseConn, error) {
-			return nil, terror.ErrDBBadConn.Generate("bad connection error restoreData")
-		},
+// `push` will append a job to the queue
+func (q *jobQueue) push(job *restoreSchemaJob) error {
+	var err error
+	select {
+	case <-q.ctx.Done():
+		err = q.ctx.Err()
+	case q.msgq <- job:
 	}
+	return terror.WithScope(err, terror.ScopeInternal)
+}
 
-	dispatchMap := make(map[string]*fileJob)
+// `close` wait jobs done and close queue forever
+func (q *jobQueue) close() error {
+	// queue is closing
+	close(q.msgq)
+	// wait until go-routines of consumption was exited
+	return q.eg.Wait()
+}
 
-	// restore db in sort
+// `startConsumers` run multiple go-routines of job consumption with user defined handler
+func (q *jobQueue) startConsumers(handler func(ctx context.Context, job *restoreSchemaJob) error) {
+	for i := 0; i < int(q.consumerCount); i++ {
+		q.eg.Go(func() error {
+			var err error
+			var session *DBConn
+		consumeLoop:
+			for {
+				select {
+				case <-q.ctx.Done():
+					err = q.ctx.Err()
+					break consumeLoop
+				case job, active := <-q.msgq:
+					if !active {
+						break consumeLoop
+					}
+					if session == nil && job.session == nil {
+						baseConn, err := job.loader.toDB.GetBaseConn(q.ctx)
```

Comment:
this `err` is shadowed and not the same one with line 1282. Could directly return here?

Reply:
Yes, it should be.